### DISCUSSION TITLE
What is the proper way to check if the remote repository is available?

### DISCUSSION THREAD

**Discussion by RasmusN** [dennisvang/tufup#118] [Q&A]
**What is the proper way to check if the remote repository is available?**

_(Yes I know, questions are supposed to be posted to stackoverflow with the "tufup" tag. However, since the tag doesn't exist it's not possible to use it unless you already are a stackoverflow guru with 1500 reputation points.)_

I tried refreshing the client and catching the DownloadHTTPError before running the update:
```python
from tufup.client import Client
from tuf.api.exceptions import DownloadHTTPError
...
client = Client(...)
try:
  client.refresh()
except DownloadHTTPError as e:
        ...
if client.check_for_updates(pre=None):
        client.download_and_apply_update(...)
...
```
but this causes an error when calling the `client.check_for_updates` method:

`RuntimeError: Cannot update timestamp after snapshot`

**System info (please complete the following information):**
 - OS: Windows
 - OS Version: 10
 - Python version: 3.11.3

**Comment by dennisvang** [comment]
@RasmusN Good question.

In the `tufup` [source][1], you'll see that `check_for_updates()` already calls `refresh()` and catches [`tuf.api.exceptions.DownloadError`][2], so you don't have to (or shouldn't?) call `refresh()` yourself.

You'll find a usage example in the `tufup-example` [source][3].

The `refresh_required` argument to `Client()` determines what happens when a download error occurs:

If `False`, the error is ignored and your app proceeds without updates, if `True`, tufup forces the app to exit (via `sys.exit()`).

If you want to handle the download error yourself, instead of automatically exiting the interpreter, you could catch the [`SystemExit`][4] from the `check_for_updates()` call.

For example (off the top of my head):

```python
...
client = Client(
    ...,
    refresh_required=True,
)
try:
    updates = client.check_for_updates()
except SystemExit:
    updates = None
    ...  # <-- do your thing here
if updates:
    client.download_and_apply_update(...)
...
```
I haven't tested this, but I suppose it should work.

[1]: https://github.com/dennisvang/tufup/blob/6d2c707adc66fd7637739f31f9859b0298a4053d/src/tufup/client.py#L154
[2]: https://github.com/theupdateframework/python-tuf/blob/a6fb0d7ddd815a6a81f44512e270c558f3169d04/tuf/api/exceptions.py#L48
[3]: https://github.com/dennisvang/tufup-example/blob/02789ce1402231908a6e01f2fe234509a23839f4/src/myapp/__init__.py#L32
[4]: https://docs.python.org/3/library/exceptions.html#SystemExit

**Comment by RasmusN** [comment]
@dennisvang Thanks this works!

Is there a specific reason why the client exists the entire process rather than raising a (custom) exception?

**Comment by dennisvang** [comment]
@RasmusN Good to hear it works.

Basically, the intention of `refresh_required` was to offer the choice between either continuing or exiting, in case of "network" errors.

The idea was to offer a simple interface, so the app creator would not need to worry about catching the right exceptions, although you could argue it's a bit restrictive. 

Another alternative was to add a user callback argument which defaults to `sys.exit`, but the option to catch the `SystemExit` was considered sufficient at the time.

[1]: https://github.com/dennisvang/tufup/blob/6d2c707adc66fd7637739f31f9859b0298a4053d/src/tufup/client.py#L34

**Comment by RasmusN** [comment]
@dennisvang Alright I understand, makes sense
